src/plugins/media/media.js

```diff
@@ -0,0 +1,261 @@
+        for (i = 0; i < elements.length; i += 1) {
+            el = elements[i];
+            delete el.mediaAutoplay;
+            delete el.mediaAutopause;
+            delete el.mediaAutostop;
+        }
+        for (i = 0; i < gcAttributes.length; i += 1) {
+            el = gcAttributes[i];
+            el.node.removeAttribute(el.attr);
+        }
+        gcAttributes = [];
+    };
+    
+    onPlay = function (event) {
+        var type = event.target.nodeName.toLowerCase();
+        document.body.classList.add("impress-media-" + type + "-playing");
+        document.body.classList.remove("impress-media-" + type + "-paused");
+    };
+
+    onPause = function (event) {
+        var type = event.target.nodeName.toLowerCase();
+        document.body.classList.add("impress-media-" + type + "-paused");
+        document.body.classList.remove("impress-media-" + type + "-playing");
+    };
+
+    onEnded = function (event) {
+        var type = event.target.nodeName.toLowerCase();
+        document.body.classList.remove("impress-media-" + type + "-playing");
+        document.body.classList.remove("impress-media-" + type + "-paused");
+    };
+    
+    removeMediaClasses = function () {
+        var type, types;
+        types = ["video", "audio"];
+        for (type in types) {
+            document.body.classList.remove("impress-media-" + types[type] + "-playing");
+            document.body.classList.remove("impress-media-" + types[type] + "-paused");
+        }
+    };
+
+    enhanceMediaNodes = function () {
+        var i, id, media, mediaElement, type;
+        
+        //gc = event.detail.api.lib.gc;
+        media = document.querySelectorAll("audio, video");
+        for (i = 0; i < media.length; i += 1) {
+            type = media[i].nodeName.toLowerCase();
+            // Set an id to identify each media node - used e.g. by the consoleMedia plugin
+            mediaElement = media[i];
+            id = mediaElement.getAttribute("id");
+            if (id === undefined || id === null) {
+                mediaElement.setAttribute("id", "media-" + type + "-" + i);
+                gcAttributes.push({"node": mediaElement, "attr": "id"});
+            }
+            gc.addEventListener(media[i], "play", onPlay);
+            gc.addEventListener(media[i], "playing", onPlay);
+            gc.addEventListener(media[i], "pause", onPause);
+            gc.addEventListener(media[i], "ended", onEnded);
+        }
+    };
+
+    enhanceMedia = function () {
+        var steps, stepElement, i;
+        enhanceMediaNodes();
+        steps = document.getElementsByClassName("step");
+        for (i = 0; i < steps.length; i += 1) {
+            stepElement = steps[i];
+            
+            // Inherit autoplay, autostop and autopause settings from root element if there is no own setting
+            if (stepElement.dataset.mediaAutoplay === undefined || stepElement.dataset.mediaAutoplay === null) {
```

Comment:
Ah. The below approach is straightforward, but makes it impossible to do the teardown. After you populate default values into the actual DOM objects themselves, you lose track of whether they were created by the user or by this code.

My preferred solution is that you don't do this at all in the init phase, and instead implement a function that returns the right value (combination of mediaElement.dataset.media*, stepElement.dataset.media*, rootElement.dataset.media* and the default) for a given media element.

Reply:
See my above comment to lines 100 to 103. (BTW is there any way in github markdown to refer to some comment in a way that turns into a link?)

Implementing a function to get this value at runtime is indeed a solution. My thought was that caching them as properties of an object would maybe speed up the move to the next step. Doing the same work over and over again (for maybe a lot of audio objects at one single step) should be slower than calculating it once.

Of course this would make a dynamic change of those attributes at runtime easier, but I don’t consider this a common scenario.

Reply:
Ah right. I was indeed reading you wrong.

The `rel` plugin keeps these in an internal data structure `startingState`. It's a list of plain JavaScript objects, not DOM. (impress.js core does the same for its list of steps and their positions.) It's not a big difference, as that internal list too has to be reset at teardown. Even then, stylistically I'd prefer you stay with the same approach here.

> Implementing a function to get this value at runtime is indeed a solution. My thought was that caching them as properties of an object would maybe speed up the move to the next step.

You're correct, of course. In this case I prefer simpler, with no caching. It's just 3 lookups for a given element, so keeping the code simpler is a benefit too.

> Of course this would make a dynamic change of those attributes at runtime easier, but I don’t consider this a common scenario.

The use case where that happens is impressionist, the wysiwyg editor project. That's where the need to use gc comes from too. While impress.js is its own standalone project, I try to keep in mind that it should work also when embedded into something like impressionist. (Without adding too much complexity or other hardship, of course.)

Reply:
> (BTW is there any way in github markdown to refer to some comment in a way that turns into a link?)

I think the date, such as "15 hours ago" is such a link.

But you don't need to, I of course know what you were referencing. You can also reference the line number where the comment is at.